### ISSUE TITLE
AS3 Spine SWC not compiling

### ISSUE THREAD

**Issue by MathiasFabian** [HaxeFoundation/haxe#6541]
Hi,

We're looking into porting our projects from AS3 to Haxe. In the first step would just like to continue target Flash - so no libs like OpenFL or targets like HTML5 need to be used in the first step. Just simple "import flash.xxx".

We got most things working but encountered a problem with the official Spine AS3 runtime swc (https://github.com/EsotericSoftware/spine-runtimes/tree/3.6/spine-as3/spine-as3-example/lib).
As soon as we include this swc into our project we get a "Information:Invalid number of type parameters for flash.Vector" compiler error. I tried to investigate the problem by checking out the library.swf and possible problems with Haxe but couldn't find anything special. The only thing that might be a bit "sketchy" is that some classes use properties like

native public var drawOrders:Vector.<Vector.<int> >;

so basically a Vector within a Vector.

Could this cause the compiler error we get?

We would like to get this working with the swc and not switch to spinehaxe or other haxelibs as they are not officially supported and the Spine runtime changes quite often (bugfixes, new features etc.).

We're using Haxe 3.4.2 to compile the project.

Any help would be much appreciated
Reto

**Comment by ibilon** [comment]
> native public var drawOrders:Vector.<Vector. >;

That line does look like the source of the error since the inner vector doesn't have a type parameter.

To be sure you'd need to make a single class as3 code with that line, compile it to a swc and see if the error reproduce.

A quick google search shows that as3 seems to allow that, but no idea what it treat those as, `Vector.<*>` maybe?

**Comment by Simn** [comment]
I didn't know AS3 allowed this. I might know how to fix this, but I don't know how to verify the fix properly.

Do you happen to know how to compile Haxe yourself?

**Comment by MathiasFabian** [comment]
There seems to be a problem with displaying the code I posted. Here it is again with proper coding tags. As you can see, there is a proper type for the inner vector too.

`native public var drawOrders:Vector.<Vector.<int> >;`

I won't find time today to investigate further, but I noticed that the error is only triggered when I import
`import spine.animation.AnimationState;`

so I guess there is something wrong with this class. I'll try to find a solution in the upcoming days - maybe just writting some sort of interface for this class with only the methods exposed that we actually need.

`public class spine.animation.AnimationState extends Object
{
  native public function AnimationState(data:spine.animation.AnimationStateData):*;
  native public var data:spine.animation.AnimationStateData;
  native public var tracks:Vector.<spine.animation.TrackEntry>;
  native internal var events:Vector.<spine.Event>;

  native public var onStart:spine.animation.Listeners;
  native public var onInterrupt:spine.animation.Listeners;
  native public var onEnd:spine.animation.Listeners;
  native public var onDispose:spine.animation.Listeners;
  native public var onComplete:spine.animation.Listeners;

  native public var onEvent:spine.animation.Listeners;
  native internal var queue:spine.animation.EventQueue;
  native internal var propertyIDs:flash.utils.Dictionary;
  native internal var mixingTo:Vector.<spine.animation.TrackEntry>;
  native internal var animationsChanged:Boolean;

  native public var timeScale:Number = 1;
  native internal var trackEntryPool:spine.Pool;
  native public function update(delta:Number):void;
  native private function updateMixingFrom(to:spine.animation.TrackEntry,delta:Number):Boolean;
  native public function apply(skeleton:spine.Skeleton):Boolean;

  native private function applyMixingFrom(to:spine.animation.TrackEntry,skeleton:spine.Skeleton,currentPose:spine.animation.MixPose):Number;
  native private function applyRotateTimeline(timeline:spine.animation.Timeline,skeleton:spine.Skeleton,time:Number,alpha:Number,pose:spine.animation.MixPose,timelinesRotation:Vector.<Number>,i:int,firstFrame:Boolean):void;
  native private function queueEvents(entry:spine.animation.TrackEntry,animationTime:Number):void;
  native public function clearTracks():void;
  native public function clearTrack(trackIndex:int):void;

  native private function setCurrent(index:int,current:spine.animation.TrackEntry,interrupt:Boolean):void;
  native public function setAnimationByName(trackIndex:int,animationName:String,loop:Boolean):spine.animation.TrackEntry;
  native public function setAnimation(trackIndex:int,animation:spine.animation.Animation,loop:Boolean):spine.animation.TrackEntry;
  native public function addAnimationByName(trackIndex:int,animationName:String,loop:Boolean,delay:Number):spine.animation.TrackEntry;
  native public function addAnimation(trackIndex:int,animation:spine.animation.Animation,loop:Boolean,delay:Number):spine.animation.TrackEntry;

  native public function setEmptyAnimation(trackIndex:int,mixDuration:Number):spine.animation.TrackEntry;
  native public function addEmptyAnimation(trackIndex:int,mixDuration:Number,delay:Number):spine.animation.TrackEntry;
  native public function setEmptyAnimations(mixDuration:Number):void;
  native private function expandToIndex(index:int):spine.animation.TrackEntry;
  native private function trackEntry(trackIndex:int,animation:spine.animation.Animation,loop:Boolean,last:spine.animation.TrackEntry):spine.animation.TrackEntry;

  native private function disposeNext(entry:spine.animation.TrackEntry):void;
  native private function _animationsChanged():void;
  native public function getCurrent(trackIndex:int):spine.animation.TrackEntry;
  native public function clearListeners():void;
  native public function clearListenerNotifications():void;

  static native public var SUBSEQUENT:int = 0;
  static native public var FIRST:int = 1;
  static native public var DIP:int = 2;
  static native public var DIP_MIX:int = 3;
  static native internal var emptyAnimation:spine.animation.Animation;
}
`

**Comment by Simn** [comment]
I can't immediately see what would cause issues in that class. Maybe it's something this class depends on?

How do I reproduce this?

**Comment by MathiasFabian** [comment]
After some debugging I found that the problem was in a related class (Pool.as) which had the following code:

```
	public function freeAll (items:Vector):void {
		for (var i:int = 0; i < items.length; i++) {
			free(items[i]);
		}
	}
```

So the compiler error in Haxe made sense and it was a simple mistake in the Spine AS3 source. I guess the problem this was working in AS3 is that the DCE got rid of this function as it is never used. But Haxe doesn't support DCE for SWCs and thus run into this problem when compiling.

Thanks for all your help looking into this issue!

**Comment by Simn** [comment]
I suspect that this can be fixed by adding `pdyn := true;` here: https://github.com/HaxeFoundation/haxe/blob/development/src/typing/swfLoader.ml#L44

That's what it does for Array as well.

**Comment by Simn** [comment]
Moving towards Haxe 4, we will no longer look into fixing old Flash/SWF issues. We are still going to keep the generator up to date with all new Haxe features, but we don't have the resources to work on flash-specific bugs. Pull requests to fix them are still going to be reviewed and possibly accepted.
